DMA2D artifacts

papadkostas opened the pull request:
After long time found out that I D caches on the Cortex M7 affect the graphics.
In case a user needs to enable them, they should properly configure the MPU.

papadkostas committed:
Quick fix for DMA2D

Cortex M7 core I & D caches, affect the graphical performance.
For more information check Application Note AN4861, Section 4.6.2

papadkostas committed:
Enable DMA2D

Use DMA2D(ART accelerator) since the issue with artifacts have been found.

papadkostas committed:
Updated submodules to latest.

kisvegabor commented:
Great! Thanks for sharing this!
Won't disabling the cache cause huge performance loss?

papadkostas commented:
Indeed, it does affect the performance a lot, which is something I didnt notice using the theme.
Running the benchmark gave me those results bellow.
Left CPU results, middle DMA2D, right DMA2D with I&D cache off
Plain benchmark                                                      73ms   40ms     56ms
Wallpaper ON                                                          75ms   43ms    103ms
Wallpaper ON + Wp recolor                                   104ms   80ms   360ms
Wallpaper ON + Wp recolor + Shadow                  111ms   87ms    387ms
Wallpaper ON + Wp recolor + Shadow + Opacity  113ms   89ms   389ms
Will dig into it further!

papadkostas committed:
Revert "Quick fix for DMA2D"

This reverts commit 9be36b2cf4f20db91cb6e21c14da01a33c5e6533.

papadkostas committed:
Update converted project files to STM32CubeIDE

papadkostas committed:
Fix for DMA2D artifacts.
Configure MPU for SDRAM as write through and perform cache maintanance each time we update the fb.

papadkostas commented:
@kisvegabor  Please check this one.
Now the performance is much better.
Plain benchmark 39ms
Wallpaper ON 41ms
Wallpaper ON + Wp recolor 80ms
Wallpaper ON + Wp recolor + Shadow 88ms
Wallpaper ON + Wp recolor + Shadow + Opacity 89ms

embeddedt commented:
In another, unrelated project with STM32F7 I found that having write-through D-cache enabled with a framebuffer still led to graphical artifacts.

Can we just turn off caching altogether for that MPU region? It shouldn't affect performance as we almost never read from that section of memory.

Also, if we are going to be updating this project to STM32CubeIDE, there is no real need for the separate Atollic project to exist anymore, correct?

kisvegabor commented:
> In another, unrelated project with STM32F7 I found that having write-through D-cache enabled with a framebuffer still led to graphical artifacts.

It's very strange that there were artifacts even in write-through mode. 

> Can we just turn off caching altogether for that MPU region? It shouldn't affect performance as we almost never read from that section of memory.

I don't know what caching option there are in F7 MPUs but if it's possible then it sounds promising.

> Also, if we are going to be updating this project to STM32CubeIDE, there is no real need for the separate Atollic project to exist anymore, correct?

I've already heard good things about CubeIDE (I still use Ac6) and probably it'd be worth to update to it as it's the current recommended ST IDE. In this case Atollic project can be archived.

embeddedt commented:
> It's very strange that there were artifacts even in write-through mode.

I agree. I didn't have a chance to investigate further, but I'm suspicious that it may have been related to [errata 1259864](https://static.docs.arm.com/epm064408/80/Cortex-M7_Software_Developers_Errata_Notice_v8.pdf#page=11).

> I don't know what caching option there are in F7 MPUs but if it's possible then it sounds promising.

As far as I know there are options for write-back, write-through, and no caching. I will double check the Cortex-M7 reference manual but it should mainly involve changing the `IsCacheable` property for the MPU region @papadkostas added to `MPU_ACCESS_NOT_CACHEABLE`.

kisvegabor commented:
> As far as I know there are options for write-back, write-through, and no caching. I will double check the Cortex-M7 reference manual but it should mainly involve changing the IsCacheable property for the MPU region @papadkostas added to MPU_ACCESS_NOT_CACHEABLE.

Maybe easier to try it :slightly_smiling_face: 
I'd try it (at least I'd try the CubeIDE too) I don't have that board with moe now.

embeddedt commented:
> Maybe easier to try it :slightly_smiling_face:

@papadkostas will have to try it, because I don't have the F769 (only the F746).

papadkostas committed:
Reconfigure the MPU for SDRAM, update lv_conf.h to latest.

papadkostas commented:
@embeddedt thanks! just tried it and pushed the changes.
The performance haven't affected at all.

embeddedt commented on hal_stm_lvgl/tft/tft.c:
```diff
@@ -115,6 +117,9 @@ void tft_init(void)
 {
+	SCB_InvalidateICache();
+
+	SCB_CleanInvalidateDCache();
```
If I'm thinking about it correctly then it should be enough to simply clean the D-cache, right? We never need to read from the framebuffer so invalidating shouldn't be needed. Also, as far as I know, the I-Cache doesn't play a role in DMA transfers.

papadkostas replied:
In case I keep just SCB_CleanInvalidateDCache the performance is reduced, based on the benchmark without wallpaper,wp recolor, shadows and opacity, rendering takes 47ms while with SCB_InvalidateICache takes 39ms.

embeddedt replied:
Odd. I would expect an I-cache invalidation to slow it down, not speed it up.

papadkostas replied:
@embeddedt see 0:55
https://www.youtube.com/watch?v=bytXsh_aCo4

papadkostas replied:
And look at this one(not mine) F746 chip and getting much better frame rate. I guess it's because the current driver does not have the implementation for the fill and blend function of the ART accelerator, right?

kisvegabor commented:
Thank you for the tests and fixes. It can be merged from my side. 
@embeddedt ? (Note the update to CubeIDE)

embeddedt commented:
Merged. It appears that we actually only have [one Atollic project](https://github.com/littlevgl/lv_port_stm32f746_disco_atollic) (for STM32F746). I will archive that now and add a note that STM32CubeIDE is the way forward.

kisvegabor commented:
> I will archive that now and add a note that STM32CubeIDE is the way forward.

Great! Thanks!

papadkostas commented:
should I implement the functions gpu_fill_cb fill and gpu_blend_cb as well?

kisvegabor commented:
> should I implement the functions gpu_fill_cb fill and gpu_blend_cb as well?

It'd be great!

papadkostas commented:
[Not working as I expected](https://github.com/papadkostas/lv_port_stm32f769_disco_sw4stm32/commit/a19a9449eccd86738267bba59133254eaaf00b91),should i open a PR for discussion?
[Here's](https://www.youtube.com/watch?v=u532sbAft4A&feature=youtu.be) how performing and you can see the issues as well.
Also I've noticed [something](https://www.youtube.com/watch?v=dF_iI37p1uY) when benchmarking on #2.
When pressing opacity, the opacity is not updated completly until you hit any other buton than opac.
While this is visible in the benchmark, on the demo is working fine. Only appears with DMA2D!

kisvegabor commented:
@papadkostas 
What if you set `disp_drv.gpu_blend_cb = NULL`. Is it working with `gpu_fill_cb` only?

papadkostas commented:
If i disable the blend function, the redering gets back to normal times but still when I press the shadow button it crashes the application and the opacity button creates those weird lines and snow on the buttons. [Video here](https://www.youtube.com/watch?v=oRY8ziygcI8)

kisvegabor commented:
It's strange because `gpu_fill_cb` should be used only for fully covering (opacity=100%) drawings. 

Can you give a link the current version of the driver?

papadkostas commented:
> It's strange because `gpu_fill_cb` should be used only for fully covering (opacity=100%) drawings.
> 
> Can you give a link the current version of the driver?

What do you mean by "version of the driver"?
The filling function is [here ](https://github.com/papadkostas/lv_port_stm32f769_disco_sw4stm32/blob/master/hal_stm_lvgl/tft/tft.c#L172)

kisvegabor commented:
> What do you mean by "version of the driver"?

Yes, I meant the latest version of the driver (as you linked it).

Are you sure [this line](https://github.com/papadkostas/lv_port_stm32f769_disco_sw4stm32/blob/master/hal_stm_lvgl/tft/tft.c#L182) is correct? I think instead of `TFT_HOR_RES` you should use `dest_width` because when lvgl redraws only a button the width of `dest` is only the width of the button.

Besides, does it work if use line-by-line fill instead of rectangle fill?
Similarly to [this](https://github.com/littlevgl/lv_port_stm32f429_disco_sw4stm32/blob/master/hal_stm_lvgl/tft/tft.c#L232).

papadkostas commented:
@kisvegabor Terrible mistake! Thanks for pointing this out, I'm gonna try it at home late this night and I'll update you.

papadkostas commented:
@kisvegabor But dest_width - width always equals 0, right?

kisvegabor commented:
> But dest_width - width always equals 0, right?

No, they can be definitely different. 
For example, if you press a  button `dest_width` will be the width of the button because only it is refreshed. However, when e.g. its left border is drawn `width` will equal to `border_width`.

The same thing happens if you refresh the entire screen. `dest_width` will be `hor_res` but `width` will be the size of the children being drawn.
